File: online judge.kaner murillo/c++/1048.cpp
//https://onlinejudge.org/external/10/1048.pdf
//1048
//10369 arbol de recubrimiento minimo. ->kruskal-prin-10 dias

#include<vector>
#include<queue>
#include<iostream>
#include<cstdio>
#include<string>
#include<climits>
#include<tuple>
#include<map>
#include<set>
#include<algorithm>

using namespace std;

void dijkstra(long long int city, long long int finish, int index, map<long long int, long long int>&visits, map<long long int, vector<tuple<long long int, int, int, bool>>>&graph, vector<long long int>&order, set<tuple<int, long long int, int>>&conj, vector<long long int>&res){
    
    priority_queue<tuple<int, long long int, int, vector<long long int>, int>, vector<tuple<int, long long int, int, vector<long long int>, int>>, greater<tuple<int, long long int, int, vector<long long int>, int>>>q;
    q.push(make_tuple(0, city, -1, vector<long long int>(), 1));
    vector<long long int>aux;
    vector<long long int>aux2;
    tuple<int, long long int, int>auxTuple;

    int i, coste, auxCoste, tique, auxTique, auxIndex;
    long long int auxCity;
    bool flag, fini = true;

    while(!q.empty()){

        coste = get<0>(q.top());
        city = get<1>(q.top());
        tique = get<2>(q.top());
        aux = get<3>(q.top());
        index = get<4>(q.top());

        //cout<<"city: "<<city<<" tique "<< tique<<" coste "<<coste<<" index "<<index<<endl;

        q.pop();   
        
        if(city == finish && coste < visits[finish] && order[index-1] == city && index == order.size()) visits[finish] = coste, res = aux, fini = false;
        
        //else{

            //cout<<"gou"<<endl<<endl;

            for(i = 0; i<graph[city].size(); i++){

                auxTique = get<2>(graph[city][i]);

                if((get<3>(graph[city][i])) || (tique == auxTique && !get<3>(graph[city][i]))){

                    auxCity = get<0>(graph[city][i]);
                    auxCoste = coste;
                    aux2 = aux;
                    auxIndex = index;
                    flag = false;

                    if(tique != auxTique) auxCoste+= get<1>(graph[city][i]), aux2.push_back(auxTique+1);
                    if(order[index] == auxCity) auxIndex++;

                    //if((auxCoste < visits[auxCity]) || !order[auxCity]){

                    auxTuple = make_tuple(auxIndex, auxCity, auxTique);

                    auto it = conj.find(auxTuple);
                    //for(it = conj.begin(); it!= conj.end() && (get<0>(*it) != auxIndex || get<1>(*it) != auxCity || get<2>(*it) != tique || get<3>(*it) != auxTique); it++); // implementar el estado del tiquete anterior

                    if(it == conj.end()){

                        //cout<<"aux_city: "<<auxCity<<" tique "<< auxTique<<" coste "<<auxCoste<<" index "<<auxIndex-1<<endl;

                        q.push(make_tuple(auxCoste, auxCity, auxTique, aux2, auxIndex));
                        conj.insert(auxTuple);
                        
                    }
                }
            }
            //}

            //cout<<endl<<endl;

        //}
    }
    
}


int main(){

    int offers, coste, sizeOffers, cases, offer, i, counter = 0, counter2, index;
    long long int city, pre, pos, first;

    while(scanf("%i", &offers) && offers){

        counter++;
        counter2 = 1;
        map<long long int, vector<tuple<long long int, int, int, bool>>> graph; 

        for(i = 0; i<offers; i++){

            scanf("%i %i %lld", &coste, &sizeOffers, &pre);
            sizeOffers--;
            bool flag = true;
            
            while(sizeOffers--){

                scanf("%lld", &pos);
                
                graph[pre].push_back(make_tuple(pos, coste, i, flag));
                pre = pos;
                flag = false;
            }
        }

        scanf("%i", &cases);

        while(cases--){

            map<long long int, long long int>visits;
            index = 0;
            vector<long long int>order;
            set<tuple<int, long long int, int>>conj;
            scanf("%i", &sizeOffers);
            scanf("%lld", &city);
            first = city;
            order.push_back(city);
            visits[city] = INT_MAX;
            sizeOffers--;

            while(sizeOffers--){

                scanf("%lld", &city);
                order.push_back(city);
                visits[city] = INT_MAX;
            }

            vector<long long int>res;
            dijkstra(first, city, 0, visits, graph, order, conj, res);
            printf("Case %i, Trip %i: Cost = %lld\n  Tickets used:", counter, counter2, visits[city]);
            
            for(i = 0; i<res.size(); i++){

                printf(" %lld", res[i]);
            }

            printf("\n");

            counter2++;
        }
    }

    return 0;
}

File: online judge.kaner murillo/python/1048.py
#1048
#https://onlinejudge.org/external/10/1048.pdf

from sys import stdin
from heapq import heappop, heappush

class State:

    def __init__(self, coste = 0, city = 0, index = 0, idTique = 0, indexTique = 0, res = []):

        self.coste = coste
        self.city = city
        self.index = index
        self.idTique = idTique
        self.indexTique = indexTique
        self.res =  res
    
    def __lt__(self, other):

        if self.coste == other.coste:

            return self.index > other.index
        
        return self.coste < other.coste

    def __str__(self):
        return f"(coste: {self.coste}, city: {self.city}, position tique: {self.indexTique}, index: {self.index}, idTique: {self.idTique})"
    
def dijkstra(graph, travel):

    visits = set()
    q = []

    for i in range(len(graph)):

        if(graph[i][2] == travel[1]):

            heappush(q, State(graph[i][0], travel[1], 1, i, 1, [i+1]))
            visits.add((1, 1, i))
            
    ans = State()

    while len(q):

        state = heappop(q)
        #print(state)
        
        if(state.index == len(travel)):

            ans = state 
            break
        
        else:
                
            if(state.indexTique < len(graph[state.idTique])-1):

                state.indexTique+=1

                if(state.index < len(travel) and graph[state.idTique][state.indexTique] == travel[state.index]):
                    
                    state.index+=1

                if((state.indexTique, state.index, state.idTique) not in visits):
                    
                    state.city = graph[state.idTique][state.indexTique-1] 
                    heappush(q, state)
                    visits.add((state.indexTique, state.index, state.idTique))

                    city = graph[state.idTique][state.indexTique]

                    for i in range(len(graph)):

                        if graph[i][2] == city:

                            res = list(state.res)
                            res.append(i+1)
                            heappush(q, State(state.coste + graph[i][0], city, state.index, i, 1, res))

    return ans

def main():

    ite = int(stdin.readline())
    counter1 = 1
    
    while ite:

        graph = []
        travel = []
        counter2 = 1

        for i in range(ite):

            graph.append([int(x) for x in stdin.readline().split()])
        
        cases = int(stdin.readline())

        for i in range(cases):

            travel = [int(x) for x in stdin.readline().split()]
            
            ans = dijkstra(graph, travel)
        
            print(f"Case {counter1}, Trip {counter2}: Cost = {ans.coste}")
            print("  Tickets used:", end="")

            for n in ans.res:

                print(f' {n}', end = "")
        
            print()
            counter2+=1
        
        counter1+=1
        ite = int(stdin.readline())

main()